File: tests/smoke-test-api.py
"""
Backend Smoke Test

Verifies all API endpoints respond correctly before starting SDK development.
Run with: python tests/smoke-test-api.py

UPDATE THESE VARIABLES before running:
"""
import sys
import math
import requests

BASE_URL = 'http://localhost:8000'
USERNAME = 'SDK_demo'
PASSWORD = 'demo_2026_@@'
MARKET_ID = '15'

# ─────────────────────────────────────────────

results: list[dict] = []
token: str | None = None


def test(name: str, method: str, fn):
    try:
        detail = fn()
        results.append({'endpoint': name, 'method': method, 'status': 'PASS', 'detail': detail})
    except Exception as e:
        results.append({'endpoint': name, 'method': method, 'status': 'FAIL', 'detail': str(e)})


def run():
    global token

    # 1. Auth
    def test_login():
        global token
        res = requests.post(
            f'{BASE_URL}/api/auth/login',
            json={'username': USERNAME, 'password': PASSWORD}
        )
        data = res.json()
        if not data.get('success') or not data.get('access_token'):
            raise Exception(f"Unexpected response: {data}")
        token = data['access_token']
        return f"Got token ({token[:20]}...)"

    test('/api/auth/login', 'POST', test_login)

    def headers():
        return {
            'Authorization': f'Bearer {token}',
            'Content-Type': 'application/json',
        }

    # 2. Market State
    def test_market_state():
        res = requests.get(f'{BASE_URL}/api/market/state?market_id={MARKET_ID}', headers=headers())
        data = res.json()
        fields = ['alpha_vector', 'market_params', 'title', 'total_volume', 'current_pool']
        missing = [f for f in fields if f not in data]
        if missing:
            raise Exception(f"Missing fields: {', '.join(missing)}. Got keys: {', '.join(data.keys())}")
        return f"title=\"{data['title']}\", K={data.get('market_params', {}).get('K')}, alpha_vector length={len(data.get('alpha_vector', []))}"

    test('/api/market/state', 'GET', test_market_state)

    # 3. Positions
    def test_positions():
        res = requests.get(f'{BASE_URL}/api/market/positions?market_id={MARKET_ID}', headers=headers())
        data = res.json()
        if 'positions' not in data:
            raise Exception(f"No 'positions' key. Got: {', '.join(data.keys())}")
        return f"{len(data['positions'])} positions found"

    test('/api/market/positions', 'GET', test_positions)

    # 4. Sell Simulate (use first open position if available, otherwise skip)
    def test_sell_simulate():
        pos_res = requests.get(f'{BASE_URL}/api/market/positions?market_id={MARKET_ID}', headers=headers())
        pos_data = pos_res.json()
        positions = pos_data.get('positions', [])
        open_pos = next((p for p in positions if p.get('status') == 'open'), None)
        if not open_pos:
            return 'SKIPPED — no open positions to simulate'
        res = requests.get(f"{BASE_URL}/api/sell/simulate/{open_pos['position_id']}?market_id={MARKET_ID}", headers=headers())
        data = res.json()
        if 'current_value_t_star' not in data:
            raise Exception(f"Missing current_value_t_star. Got: {', '.join(data.keys())}")
        return f"position_id={data['position_id']}, value={data['current_value_t_star']}"

    test('/api/sell/simulate/{id}', 'GET', test_sell_simulate)

    # 5. Project Settlement
    def test_project_settlement():
        state_res = requests.get(f'{BASE_URL}/api/market/state?market_id={MARKET_ID}', headers=headers())
        state_data = state_res.json()
        K = state_data['market_params']['K']
        belief = [1 / (K + 1)] * (K + 1)
        res = requests.post(
            f'{BASE_URL}/api/projection/project_settlement?market_id={MARKET_ID}',
            headers=headers(),
            json={'belief_vector': belief, 'collateral': 10, 'num_outcomes': 20}
        )
        data = res.json()
        if not data.get('projections'):
            raise Exception(f"No projections. Got: {', '.join(data.keys())}")
        return f"{len(data['projections'])} projections, max_payout={data.get('max_payout')}"

    test('/api/projection/project_settlement', 'POST', test_project_settlement)

    # 6. Buy (small trade to verify endpoint works)
    def test_buy():
        state_res = requests.get(f'{BASE_URL}/api/market/state?market_id={MARKET_ID}', headers=headers())
        state_data = state_res.json()
        K = state_data['market_params']['K']
        L = state_data['market_params']['L']
        H = state_data['market_params']['H']
        center = (L + H) / 2
        # Simple Gaussian belief
        belief = []
        for k in range(K + 1):
            u = k / K
            u_center = (center - L) / (H - L)
            belief.append(math.exp(-0.5 * ((u - u_center) / 0.15) ** 2))
        total = sum(belief)
        normalized = [v / total for v in belief]
        res = requests.post(
            f'{BASE_URL}/api/market/buy?market_id={MARKET_ID}',
            headers=headers(),
            json={'C': 1, 'p_vector': normalized, 'prediction': center}
        )
        data = res.json()
        if not data.get('success') or not data.get('position'):
            raise Exception(f"Buy failed: {data}")
        return f"position_id={data['position']['position_id']}, claims={data['position']['minted_claims_m']}"

    test('/api/market/buy', 'POST', test_buy)

    # Print results
    print('\n=== Backend Smoke Test Results ===\n')
    for r in results:
        icon = 'OK' if r['status'] == 'PASS' else 'FAIL'
        print(f"[{icon}] {r['method']} {r['endpoint']}")
        print(f"     {r['detail']}\n")

    failed = [r for r in results if r['status'] == 'FAIL']
    if failed:
        print(f"\n{len(failed)} endpoint(s) FAILED. Fix before starting development.")
        sys.exit(1)
    else:
        print('\nAll endpoints verified. Ready to build.')


if __name__ == '__main__':
    run()

File: tests/smoke-test-api.ts
/**
 * Backend Smoke Test
 *
 * Verifies all API endpoints respond correctly before starting SDK development.
 * Run with: npx tsx tests/smoke-test-api.ts
 *
 * Set these environment variables before running (or create a root .env file):
 *   FS_TEST_URL, FS_TEST_USERNAME, FS_TEST_PASSWORD, FS_TEST_MARKET_ID
 */
const BASE_URL = 'http://localhost:8000';
const USERNAME = '';
const PASSWORD =  '';
const MARKET_ID = '15';

if (!USERNAME || !PASSWORD) {
  console.error('Missing FS_TEST_USERNAME or FS_TEST_PASSWORD environment variables.');
  console.error('Set them in your shell or in a root .env file.');
  process.exit(1);
}

// ─────────────────────────────────────────────

interface TestResult {
  endpoint: string;
  method: string;
  status: 'PASS' | 'FAIL';
  detail: string;
}

const results: TestResult[] = [];
let token: string | null = null;

async function test(name: string, method: string, fn: () => Promise<string>) {
  try {
    const detail = await fn();
    results.push({ endpoint: name, method, status: 'PASS', detail });
  } catch (e: any) {
    results.push({ endpoint: name, method, status: 'FAIL', detail: e.message });
  }
}

async function run() {
  // 1. Auth
  await test('/api/auth/login', 'POST', async () => {
    const res = await fetch(`${BASE_URL}/api/auth/login`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ username: USERNAME, password: PASSWORD }),
    });
    const data = await res.json();
    if (!data.success || !data.access_token) throw new Error(`Unexpected response: ${JSON.stringify(data)}`);
    token = data.access_token;
    return `Got token (${token!.substring(0, 20)}...)`;
  });

  const headers = () => ({
    'Authorization': `Bearer ${token}`,
    'Content-Type': 'application/json',
  });

  // 2. Market State
  await test('/api/market/state', 'GET', async () => {
    const res = await fetch(`${BASE_URL}/api/market/state?market_id=${MARKET_ID}`, { headers: headers() });
    const data = await res.json();
    const fields = ['alpha_vector', 'market_params', 'title', 'total_volume', 'current_pool'];
    const missing = fields.filter(f => !(f in data));
    if (missing.length) throw new Error(`Missing fields: ${missing.join(', ')}. Got keys: ${Object.keys(data).join(', ')}`);
    return `title="${data.title}", K=${data.market_params?.K}, alpha_vector length=${data.alpha_vector?.length}`;
  });

  // 3. Positions
  await test('/api/market/positions', 'GET', async () => {
    const res = await fetch(`${BASE_URL}/api/market/positions?market_id=${MARKET_ID}`, { headers: headers() });
    const data = await res.json();
    if (!('positions' in data)) throw new Error(`No 'positions' key. Got: ${Object.keys(data).join(', ')}`);
    return `${data.positions.length} positions found`;
  });

  // 4. Sell Simulate (use first open position if available, otherwise skip)
  await test('/api/sell/simulate/{id}', 'GET', async () => {
    const posRes = await fetch(`${BASE_URL}/api/market/positions?market_id=${MARKET_ID}`, { headers: headers() });
    const posData = await posRes.json();
    const openPos = posData.positions?.find((p: any) => p.status === 'open');
    if (!openPos) return 'SKIPPED — no open positions to simulate';
    const res = await fetch(`${BASE_URL}/api/sell/simulate/${openPos.position_id}?market_id=${MARKET_ID}`, { headers: headers() });
    const data = await res.json();
    if (!('current_value_t_star' in data)) throw new Error(`Missing current_value_t_star. Got: ${Object.keys(data).join(', ')}`);
    return `position_id=${data.position_id}, value=${data.current_value_t_star}`;
  });

  // 5. Preview Settlement
  await test('/api/projection/project_settlement', 'POST', async () => {
    // Build a simple uniform belief vector
    const stateRes = await fetch(`${BASE_URL}/api/market/state?market_id=${MARKET_ID}`, { headers: headers() });
    const stateData = await stateRes.json();
    const K = stateData.market_params.K;
    const belief = Array(K + 1).fill(1 / (K + 1));
    const res = await fetch(`${BASE_URL}/api/projection/project_settlement?market_id=${MARKET_ID}`, {
      method: 'POST',
      headers: headers(),
      body: JSON.stringify({ belief_vector: belief, collateral: 10, num_outcomes: 20 }),
    });
    const data = await res.json();
    if (!data.projections) throw new Error(`No projections. Got: ${Object.keys(data).join(', ')}`);
    return `${data.projections.length} projections, max_payout=${data.max_payout}`;
  });

  // 6. Buy (small trade to verify endpoint works)
  await test('/api/market/buy', 'POST', async () => {
    const stateRes = await fetch(`${BASE_URL}/api/market/state?market_id=${MARKET_ID}`, { headers: headers() });
    const stateData = await stateRes.json();
    const K = stateData.market_params.K;
    const L = stateData.market_params.L;
    const H = stateData.market_params.H;
    const center = (L + H) / 2;
    // Simple Gaussian belief
    const belief = Array.from({ length: K + 1 }, (_, k) => {
      const u = k / K;
      const uCenter = (center - L) / (H - L);
      return Math.exp(-0.5 * ((u - uCenter) / 0.15) ** 2);
    });
    const sum = belief.reduce((a, b) => a + b, 0);
    const normalized = belief.map(v => v / sum);
    const res = await fetch(`${BASE_URL}/api/market/buy?market_id=${MARKET_ID}`, {
      method: 'POST',
      headers: headers(),
      body: JSON.stringify({ C: 1, p_vector: normalized, prediction: center }),
    });
    const data = await res.json();
    if (!data.success || !data.position) throw new Error(`Buy failed: ${JSON.stringify(data)}`);
    return `position_id=${data.position.position_id}, claims=${data.position.minted_claims_m}`;
  });

  // Print results
  console.log('\n=== Backend Smoke Test Results ===\n');
  for (const r of results) {
    const icon = r.status === 'PASS' ? 'OK' : 'FAIL';
    console.log(`[${icon}] ${r.method} ${r.endpoint}`);
    console.log(`     ${r.detail}\n`);
  }
  const failed = results.filter(r => r.status === 'FAIL');
  if (failed.length) {
    console.log(`\n${failed.length} endpoint(s) FAILED. Fix before starting development.`);
    process.exit(1);
  } else {
    console.log('\nAll endpoints verified. Ready to build.');
  }
}

run();
